File: iosApp/iosApp/Services/AuthStateManager.swift
//
//  AuthStateManager.swift
//  iosApp
//
//  Created by Wakev on 2025
//

import Foundation
import Combine

/**
 * Authentication state enum.
 */
enum AuthState {
    case loading
    case unauthenticated
    case authenticated(userId: String, user: UserResponse)
    case error(message: String)
}

/**
 * Central authentication state manager for iOS.
 *
 * This class manages the authentication state across the entire application using
 * Combine publishers for reactive updates. It handles:
 * - Initial authentication check on app startup
 * - OAuth login flow coordination
 * - Token refresh and expiry handling
 * - Logout and session cleanup
 */
@MainActor
class AuthStateManager: ObservableObject {
    @Published private(set) var authState: AuthState = .loading
    @Published private(set) var currentUserId: String?
    @Published private(set) var currentAccessToken: String?

    private let authService: AuthenticationService
    private let enableOAuth: Bool
    private var tokenRefreshTask: Task<Void, Never>?

    init(
        authService: AuthenticationService,
        enableOAuth: Bool = false
    ) {
        self.authService = authService
        self.enableOAuth = enableOAuth
    }

    /**
     * Initialize authentication state.
     *
     * This should be called when the app starts. It checks for existing
     * authentication and validates the stored tokens.
     */
    func initialize() async {
        authState = .loading

        // If OAuth is disabled, skip auth check
        guard enableOAuth else {
            authState = .unauthenticated
            return
        }

        do {
            // Check if user is already logged in
            let isAuthenticated = await authService.isAuthenticated()

            if isAuthenticated {
                // Get user information
                guard let userId = await authService.getCurrentUserId(),
                      let accessToken = await authService.getAccessToken() else {
                    await logout()
                    return
                }

                currentUserId = userId
                currentAccessToken = accessToken

                // Create user response
                // TODO: Fetch full user profile from server
                let user = UserResponse(
                    id: userId,
                    email: "user@example.com",
                    name: "User",
                    avatarUrl: nil,
                    provider: "apple",
                    createdAt: ISO8601DateFormatter().string(from: Date())
                )

                authState = .authenticated(userId: userId, user: user)

                // Start token refresh monitoring
                startTokenRefreshMonitoring()
            } else {
                authState = .unauthenticated
            }
        } catch {
            authState = .error(message: "Failed to initialize authentication: \(error.localizedDescription)")
        }
    }

    /**
     * Login with OAuth provider.
     */
    func login(provider: String, authCode: String, userInfo: String? = nil) async {
        authState = .loading

        do {
            let response: OAuthLoginResponse

            if provider == "apple" {
                response = try await authService.loginWithApple(
                    authorizationCode: authCode,
                    userInfo: userInfo
                )
            } else {
                response = try await authService.loginWithGoogle(
                    authorizationCode: authCode
                )
            }

            currentUserId = response.user.id
            currentAccessToken = response.accessToken
            authState = .authenticated(userId: response.user.id, user: response.user)

            // Start token refresh monitoring
            startTokenRefreshMonitoring()

        } catch {
            authState = .error(message: "Login failed: \(error.localizedDescription)")
        }
    }

    /**
     * Logout the current user.
     */
    func logout() async {
        // Stop token refresh monitoring
        tokenRefreshTask?.cancel()
        tokenRefreshTask = nil

        // Clear tokens
        try? await authService.logout()

        // Update state
        currentUserId = nil
        currentAccessToken = nil
        authState = .unauthenticated
    }

    /**
     * Refresh access token if needed.
     */
    func refreshTokenIfNeeded() async {
        do {
            let response = try await authService.refreshToken()
            currentAccessToken = response.accessToken

            // Update authenticated state with new token
            if case .authenticated(let userId, _) = authState {
                authState = .authenticated(userId: userId, user: response.user)
            }
        } catch {
            // Refresh failed - logout user
            await logout()
        }
    }

    /**
     * Start background token refresh monitoring.
     *
     * Checks token expiry periodically and refreshes proactively.
     */
    private func startTokenRefreshMonitoring() {
        tokenRefreshTask?.cancel()
        tokenRefreshTask = Task {
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 5 * 60 * 1_000_000_000) // 5 minutes

                if case .authenticated = authState {
                    await refreshTokenIfNeeded()
                }
            }
        }
    }

    #if DEBUG
    /**
     * Set authentication state for development mode.
     * Bypasses normal OAuth flow.
     */
    func setAuthStateForDevelopment(userId: String, accessToken: String) {
        currentUserId = userId
        currentAccessToken = accessToken

        // Create mock user
        let mockUser = UserResponse(
            id: userId,
            email: "dev@wakeve.local",
            name: "Dev User",
            avatarUrl: nil,
            provider: "development",
            createdAt: ISO8601DateFormatter().string(from: Date())
        )

        authState = .authenticated(userId: userId, user: mockUser)

        print("✅ Development mode: Authenticated as \(userId)")
    }
    #endif

    /**
     * Clean up resources.
     */
    func dispose() {
        tokenRefreshTask?.cancel()
        tokenRefreshTask = nil
    }
}

File: shared/src/commonMain/kotlin/com/guyghost/wakeve/auth/AuthStateManager.kt
package com.guyghost.wakeve.auth

import com.guyghost.wakeve.currentTimeMillis
import com.guyghost.wakeve.models.OAuthProvider
import com.guyghost.wakeve.models.UserResponse
import com.guyghost.wakeve.security.SecureTokenStorage
import com.guyghost.wakeve.security.UserProfileData
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

/**
 * Authentication state sealed class.
 *
 * Represents all possible authentication states in the application.
 */
sealed class AuthState {
    /**
     * Initial state while checking for existing authentication.
     */
    object Loading : AuthState()

    /**
     * User is not authenticated.
     */
    object Unauthenticated : AuthState()

    /**
     * User is successfully authenticated.
     *
     * @property userId The unique user identifier
     * @property user The user profile information
     * @property sessionId The current session identifier
     */
    data class Authenticated(
        val userId: String,
        val user: UserResponse,
        val sessionId: String
    ) : AuthState()

    /**
     * Authentication error occurred.
     *
     * @property message User-friendly error message
     * @property code Error code for debugging
     */
    data class Error(
        val message: String,
        val code: ErrorCode = ErrorCode.UNKNOWN
    ) : AuthState()
}

/**
 * Error codes for authentication failures.
 */
enum class ErrorCode {
    UNKNOWN,
    NETWORK_ERROR,
    INVALID_CREDENTIALS,
    TOKEN_EXPIRED,
    SERVER_ERROR,
    USER_CANCELLED
}

/**
 * Central authentication state manager.
 *
 * This class manages the authentication state across the entire application using
 * StateFlow for reactive updates. It handles:
 * - Initial authentication check on app startup
 * - OAuth login flow coordination
 * - Token refresh and expiry handling
 * - Logout and session cleanup
 * - Feature flag integration for progressive rollout
 *
 * Usage:
 * ```kotlin
 * val authStateManager = AuthStateManager(secureStorage, authService)
 * val authState by authStateManager.authState.collectAsState()
 *
 * when (authState) {
 *     is AuthState.Loading -> LoadingScreen()
 *     is AuthState.Unauthenticated -> LoginScreen()
 *     is AuthState.Authenticated -> MainApp()
 *     is AuthState.Error -> ErrorScreen()
 * }
 * ```
 */
class AuthStateManager(
    private val secureStorage: SecureTokenStorage,
    private val authService: ClientAuthenticationService,
    private val enableOAuth: Boolean = false // Feature flag
) {
    private val _authState = MutableStateFlow<AuthState>(AuthState.Loading)
    val authState: StateFlow<AuthState> = _authState

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private var tokenRefreshJob: Job? = null

    init {
        scope.launch {
            checkExistingAuth()
        }
    }

    /**
     * Check for existing authentication on startup.
     */
    private suspend fun checkExistingAuth() {
        val accessToken = secureStorage.getAccessToken()
        val userId = secureStorage.getUserId()

        if (accessToken != null && userId != null && !secureStorage.isTokenExpired()) {
            // User is logged in, fetch full profile
            fetchAndSetUserProfile(userId)
        } else {
            _authState.value = AuthState.Unauthenticated
        }
    }

    /**
     * Fetch full user profile from server.
     */
    private suspend fun fetchAndSetUserProfile(userId: String) {
        _authState.value = AuthState.Loading

        try {
            val response = authService.getUserProfile(userId)
            if (response.isSuccess && response.valueOrNull() != null) {
                val userProfile = response.valueOrNull()!!

                // Store user profile data
                secureStorage.storeUserName(userProfile.name)
                secureStorage.storeUserEmail(userProfile.email)
                secureStorage.storeUserProvider(userProfile.provider.toString())

                // Set authenticated state
                val session = generateSessionId()
                secureStorage.storeSessionId(session)

                _authState.value = AuthState.Authenticated(
                    userId = userId,
                    user = userProfile,
                    sessionId = session
                )
            } else {
                _authState.value = AuthState.Error(
                    message = "Failed to fetch user profile",
                    code = ErrorCode.SERVER_ERROR
                )
            }
        } catch (e: Exception) {
            _authState.value = AuthState.Error(
                message = e.message ?: "Unknown error",
                code = mapErrorCode(e)
            )
        }
    }

    /**
     * Login with Google OAuth.
     */
    suspend fun loginWithGoogle(authCode: String) {
        if (!enableOAuth) {
            _authState.value = AuthState.Error(
                message = "OAuth is not enabled",
                code = ErrorCode.UNKNOWN
            )
            return
        }

        _authState.value = AuthState.Loading

        try {
            val response = authService.loginWithGoogle(authCode)

            if (response.isSuccess && response.valueOrNull() != null) {
                val loginResponse = response.valueOrNull()!!

                // Store tokens
                secureStorage.storeAccessToken(loginResponse.accessToken)
                secureStorage.storeRefreshToken(loginResponse.refreshToken)
                secureStorage.storeTokenExpiry(loginResponse.expiresIn * 1000L + currentTimeMillis())

                // Store user profile
                secureStorage.storeUserId(loginResponse.user.id)
                secureStorage.storeUserName(loginResponse.user.name)
                secureStorage.storeUserEmail(loginResponse.user.email)
                secureStorage.storeUserProvider(loginResponse.user.provider.toString())

                // Set authenticated state
                val session = generateSessionId()
                secureStorage.storeSessionId(session)

                _authState.value = AuthState.Authenticated(
                    userId = loginResponse.user.id,
                    user = loginResponse.user,
                    sessionId = session
                )
            } else {
                _authState.value = AuthState.Error(
                    message = "OAuth login failed",
                    code = ErrorCode.INVALID_CREDENTIALS
                )
            }
        } catch (e: Exception) {
            _authState.value = AuthState.Error(
                message = e.message ?: "Unknown error",
                code = mapErrorCode(e)
            )
        }
    }

    /**
     * Login with Apple Sign In.
     */
    suspend fun loginWithApple(authCode: String, userInfo: String? = null) {
        if (!enableOAuth) {
            _authState.value = AuthState.Error(
                message = "OAuth is not enabled",
                code = ErrorCode.UNKNOWN
            )
            return
        }

        _authState.value = AuthState.Loading

        try {
            val response = authService.loginWithApple(authCode, userInfo)

            if (response.isSuccess && response.valueOrNull() != null) {
                val loginResponse = response.valueOrNull()!!

                // Store tokens
                secureStorage.storeAccessToken(loginResponse.accessToken)
                secureStorage.storeRefreshToken(loginResponse.refreshToken)
                secureStorage.storeTokenExpiry(loginResponse.expiresIn * 1000L + currentTimeMillis())

                // Store user profile
                secureStorage.storeUserId(loginResponse.user.id)
                secureStorage.storeUserName(loginResponse.user.name)
                secureStorage.storeUserEmail(loginResponse.user.email)
                secureStorage.storeUserProvider(loginResponse.user.provider.toString())

                // Set authenticated state
                val session = generateSessionId()
                secureStorage.storeSessionId(session)

                _authState.value = AuthState.Authenticated(
                    userId = loginResponse.user.id,
                    user = loginResponse.user,
                    sessionId = session
                )
            } else {
                _authState.value = AuthState.Error(
                    message = "Apple login failed",
                    code = ErrorCode.INVALID_CREDENTIALS
                )
            }
        } catch (e: Exception) {
            _authState.value = AuthState.Error(
                message = e.message ?: "Unknown error",
                code = mapErrorCode(e)
            )
        }
    }

    /**
     * Refresh access token.
     */
    suspend fun refreshToken() {
        _authState.value = AuthState.Loading

        try {
            val response = authService.refreshToken()

            if (response.isSuccess && response.valueOrNull() != null) {
                val refreshResponse = response.valueOrNull()!!

                // Update tokens
                secureStorage.storeAccessToken(refreshResponse.accessToken)
                secureStorage.storeRefreshToken(refreshResponse.refreshToken)
                secureStorage.storeTokenExpiry(refreshResponse.expiresIn * 1000L + currentTimeMillis())

                _authState.value = AuthState.Authenticated(
                    userId = getCurrentUserId(),
                    user = getCurrentUser(),
                    sessionId = getCurrentSessionId()
                )
            } else {
                _authState.value = AuthState.Error(
                    message = "Token refresh failed",
                    code = ErrorCode.TOKEN_EXPIRED
                )
            }
        } catch (e: Exception) {
            _authState.value = AuthState.Error(
                message = e.message ?: "Unknown error",
                code = mapErrorCode(e)
            )
        }
    }

    /**
     * Logout user.
     */
    suspend fun logout() {
        _authState.value = AuthState.Loading

        try {
            authService.logout()
            secureStorage.clearAllTokens()
            _authState.value = AuthState.Unauthenticated
        } catch (e: Exception) {
            _authState.value = AuthState.Error(
                message = e.message ?: "Logout failed",
                code = mapErrorCode(e)
            )
        }
    }

    /**
     * Check if user is logged in.
     */
    suspend fun isLoggedIn(): Boolean {
        return authState.value is AuthState.Authenticated
    }

    /**
     * Get current user ID.
     */
    suspend fun getCurrentUserId(): String? {
        return secureStorage.getUserId()
    }

    /**
     * Get current user.
     */
    suspend fun getCurrentUser(): UserResponse? {
        val userId = secureStorage.getUserId()
        val email = secureStorage.getUserEmail()
        val name = secureStorage.getUserName()
        val providerStr = secureStorage.getUserProvider()

        if (userId != null && email != null && name != null && providerStr != null) {
            val provider = try {
                OAuthProvider.valueOf(providerStr)
            } catch (e: Exception) {
                OAuthProvider.GOOGLE
            }

            return UserResponse(
                id = userId,
                email = email,
                name = name,
                provider = provider,
                profilePictureUrl = secureStorage.getUserAvatarUrl()
            )
        }

        return null
    }

    /**
     * Get current session ID.
     */
    suspend fun getCurrentSessionId(): String? {
        return secureStorage.getSessionId()
    }

    /**
     * Map exception to error code.
     */
    private fun mapErrorCode(error: Throwable): ErrorCode {
        return when {
            error.message?.contains("network", ignoreCase = true) == true -> ErrorCode.NETWORK_ERROR
            error.message?.contains("invalid", ignoreCase = true) == true -> ErrorCode.INVALID_CREDENTIALS
            error.message?.contains("expired", ignoreCase = true) == true -> ErrorCode.TOKEN_EXPIRED
            error.message?.contains("server", ignoreCase = true) == true -> ErrorCode.SERVER_ERROR
            error.message?.contains("cancel", ignoreCase = true) == true -> ErrorCode.USER_CANCELLED
            else -> ErrorCode.UNKNOWN
        }
    }

    /**
     * Get current timestamp in ISO 8601 format.
     */
    private fun getCurrentTimestamp(): String {
        // Simple ISO 8601 timestamp
        return "${currentTimeMillis()}"
    }

    /**
     * Clean up resources.
     */
    fun dispose() {
        tokenRefreshJob?.cancel()
        scope.cancel()
    }

    companion object {
        /**
         * Singleton instance for simple access.
         *
         * Note: For production use, consider dependency injection (Koin).
         *
         * @param secureStorage Optional secure storage (lazy initialized if not provided)
         * @param authService Optional auth service (lazy initialized if not provided)
         * @param enableOAuth Feature flag for OAuth
         */
        private var instance: AuthStateManager? = null

        /**
         * Get singleton instance of AuthStateManager.
         *
         * Note: For production use, consider dependency injection (Koin).
         *
         * @param secureStorage Optional secure storage (lazy initialized if not provided)
         * @param authService Optional auth service (lazy initialized if not provided)
         * @param enableOAuth Feature flag for OAuth
         */
        fun getInstance(
            secureStorage: SecureTokenStorage? = null,
            authService: ClientAuthenticationService? = null,
            enableOAuth: Boolean = false
        ): AuthStateManager {
            // Use memoization for thread-safe lazy initialization
            val currentValue = instance

            if (currentValue != null) {
                return currentValue
            }

            val newInstance = AuthStateManager(
                secureStorage = secureStorage ?: DummySecureTokenStorage(),
                authService = authService ?: DummyAuthenticationService(),
                enableOAuth = enableOAuth
            )

            instance = newInstance
            return newInstance
        }

        /**
         * Check if instance exists
         */
        fun hasInstance(): Boolean = instance != null

        /**
         * Clear the singleton instance (for testing)
         */
        fun clearInstance() {
            instance?.dispose()
            instance = null
        }
    }
}

/**
 * Generate a unique session ID.
 */
private fun generateSessionId(): String {
    return "session-${currentTimeMillis()}-${(0..999).random()}"
}

/**
 * Simple authentication error class.
 */
private class NotImplementedError(message: String) : Exception(message)

/**
 * Dummy implementation of SecureTokenStorage for shared module compilation.
 */
private class DummySecureTokenStorage : SecureTokenStorage {
    override suspend fun storeAccessToken(token: String) = Result.success(Unit)
    override suspend fun storeRefreshToken(token: String) = Result.success(Unit)
    override suspend fun storeUserId(userId: String) = Result.success(Unit)
    override suspend fun storeTokenExpiry(expiryTimestamp: Long) = Result.success(Unit)
    override suspend fun storeUserEmail(email: String) = Result.success(Unit)
    override suspend fun storeUserName(name: String) = Result.success(Unit)
    override suspend fun storeUserProvider(provider: String) = Result.success(Unit)
    override suspend fun storeUserAvatarUrl(avatarUrl: String?) = Result.success(Unit)
    override suspend fun storeUserProfile(profile: UserProfileData) = Result.success(Unit)
    override suspend fun getAccessToken(): String? = null
    override suspend fun getRefreshToken(): String? = null
    override suspend fun getUserId(): String? = null
    override suspend fun getTokenExpiry(): Long? = null
    override suspend fun getUserEmail(): String? = null
    override suspend fun getUserName(): String? = null
    override suspend fun getUserProvider(): String? = null
    override suspend fun getUserAvatarUrl(): String? = null
    override suspend fun getUserProfile(): UserProfileData? = null
    override suspend fun getSessionId(): String? = null
    override suspend fun storeSessionId(sessionId: String) = Result.success(Unit)
    override suspend fun clearAllTokens() = Result.success(Unit)
    override suspend fun isTokenExpired(): Boolean = true
    override suspend fun hasValidToken(): Boolean = false
}

/**
 * Dummy implementation of ClientAuthenticationService for shared module compilation.
 */
private class DummyAuthenticationService : ClientAuthenticationService {
    override suspend fun loginWithGoogle(authCode: String): Result<AuthResult<UserResponse>> =
        Result.failure(NotImplementedError("Use platform-specific implementation"))

    override suspend fun loginWithApple(authCode: String, userInfo: String?): Result<AuthResult<UserResponse>> =
        Result.failure(NotImplementedError("Use platform-specific implementation"))

    override suspend fun refreshToken(): Result<AuthResult<UserResponse>> =
        Result.failure(NotImplementedError("Use platform-specific implementation"))

    override suspend fun logout(): Result<Unit> = Result.success(Unit)

    override fun isLoggedIn(): Boolean = false
}

/**
 * Result wrapper for auth operations.
 */
sealed class AuthResult<out T> {
    data class Success<out T>(val value: T) : AuthResult<T>()
    data class Failure(val error: Exception) : AuthResult<Nothing>()

    val isSuccess: Boolean get() = this is Success
    val isFailure: Boolean get() = this is Failure

    fun valueOrNull(): T? = when (this) {
        is Success<T> -> value
        is Failure -> null
    }
}
